Webfonts: Add internal-only theme.json webfonts handler - WP 6.0 stopgap

hellofromtonya opened the pull request:
Part of #40472.

Props to @aristath who pulled together the key elements of this PR.

<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Using the Webfonts API from PR #37140, this PR adds a `theme.json` webfonts handler (albeit funky design) to:

* handle only `theme.json` webfonts defined in `fontFace`.
* prevent access to the webfonts handling mechanisms by encapsulating all logic into one function that serves as a private container, i.e. `_wp_theme_json_webfonts_handler()`.

This is a stopgap approach.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
See the issue for the reasons.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

It uses the Webfonts API from PR #37140 but smushes the code together into `$fn = function() {}` closures that are nested inside of a global function. This global function is the container that prevents access to the handler, thus ensuring the mechanisms are hidden and not accessible for extenders to interact with the handler.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->

#### Set up instructions:

Using Twenty Twenty-Two theme, either:
* Option 1: Download Twenty Twenty-Two's new style variations https://github.com/WordPress/wordpress-develop/pull/2440
* Option 2: Modify the current version in WordPress Core by:
     * In its `functions.php` file, comment out lines 58 and 81 (the lines that are add inline styles using the font face styles).
     * Replace the `fontFamilies` in its `theme.json` file

```
			"fontFamilies": [
				{
					"fontFamily": "-apple-system,BlinkMacSystemFont,\"Segoe UI\",Roboto,Oxygen-Sans,Ubuntu,Cantarell,\"Helvetica Neue\",sans-serif",
					"name": "System Font",
					"slug": "system-font"
				},
				{
					"fontFamily": "\"Source Serif Pero\", serif",
					"name": "Source Serif Pero",
					"slug": "source-serif-pero",
					"fontFace": [
						{
							"fontFamily": "Source Serif Pero",
							"fontWeight": "200 900",
							"fontStyle": "normal",
							"fontStretch": "normal",
							"src": [ "file:./assets/fonts/SourceSerif4Variable-Roman.ttf.woff2" ]
						},
						{
							"fontFamily": "Source Serif Pero",
							"fontWeight": "200 900",
							"fontStyle": "italic",
							"fontStretch": "normal",
							"src": [ "file:./assets/fonts/SourceSerif4Variable-Italic.ttf.woff2" ]
						}
					]
				}
			],
```

#### How to Test

#### How to Test on the front-end

1. Activate TT2
2. Load the site's home page in the browser.
3. In the browser, right click to open the dev tools UI and then select "Inspect" menu item.
4. In the HTML inspector, open the `<head>` element to view all of the nodes within the head.
5. Search for the `<style>` element that has an ID of `wp-webfonts-inline-css`.
6. Open that element.

Expected Results:
The `<head>` should have one instance of the following HTML style element:

```hmtl
<style id="wp-webfonts-inline-css">
@font-face{font-family:"Source Serif Pro";font-style:normal;font-weight:200 900;font-display:fallback;src:local("Source Serif Pro"), url('/wp-content/themes/twentytwentytwo/assets/fonts/source-serif-pro/SourceSerif4Variable-Roman.ttf.woff2') format('woff2');font-stretch:normal;}
@font-face{font-family:"Source Serif Pro";font-style:italic;font-weight:200 900;font-display:fallback;src:local("Source Serif Pro"), url('/wp-content/themes/twentytwentytwo/assets/fonts/source-serif-pro/SourceSerif4Variable-Italic.ttf.woff2') format('woff2');font-stretch:normal;}
</style>
```

This element should have 
* 2 `@font-face` declarations to match what was defined in the `theme.json` file.
* The `url` should no longer have the `file:./` placeholder in it.

#### How to Test in the back-end

1. Go to "Appearance" > Themes and make sure TT2 is activated.
2. Go to "Appearance" > Editor (beta) to load the site editor.
3. In the browser, right click to open the dev tools UI and then select "Inspect" menu item.
4. In the HTML inspector, open the `<head>` element to view all of the nodes within the head.
5. Search for the `<style>` element that has an ID of `wp-block-library-inline-css`. Open that element (see Expected Results for the CSS).
6. In the Network tab, select "Disable Cache" and "Fonts".
7. Refresh the page. Notice the font file(s) that are loaded.

Expected Results:
The `<head>` should have one instance of the following HTML style element:

```hmtl
<style id="wp-block-library-inline-css">
@font-face{font-family:"Source Serif Pro";font-style:normal;font-weight:200 900;font-display:fallback;src:local("Source Serif Pro"), url('/wp-content/themes/twentytwentytwo/assets/fonts/source-serif-pro/SourceSerif4Variable-Roman.ttf.woff2') format('woff2');font-stretch:normal;}@font-face{font-family:"Source Serif Pro";font-style:italic;font-weight:200 900;font-display:fallback;src:local("Source Serif Pro"), url('/wp-content/themes/twentytwentytwo/assets/fonts/source-serif-pro/SourceSerif4Variable-Italic.ttf.woff2') format('woff2');font-stretch:normal;}
</style>
```

This element should have 
* 2 `@font-face` declarations to match what was defined in the `theme.json` file.
* The `url` should no longer have the `file:./` placeholder in it.

## Design Considerations

* To enable testing, this PR comments out loading the Webfonts API files in the `lib/load.php` file. In doing so, this makes the PHPUnit tests fail as those files are missing.

* A mechanism is needed in Gutenberg to continue developing, testing, and maturing the Webfonts API. 
    * This PR currently hooks  `_wp_theme_json_webfonts_handler()` into `'wp_plugins'` action event, giving Gutenberg's plugin the opportunity to remove that callback and load its files.
    * Another way to accomplish this is to create a short-circuit within the container. In doing that, the container could be an unnamed static closure registered to the `'wp_plugins'` action event. Doing this would eliminate deprecating the function once the Webfonts API is merged into Core.

hellofromtonya committed:
Adds stopgap handler for theme.json webfonts.

hellofromtonya commented on lib/load.php:
```diff
@@ -114,12 +114,18 @@ function gutenberg_is_experiment_enabled( $name ) {
-require __DIR__ . '/experimental/register-webfonts-from-theme-json.php';
+require __DIR__ . '/experimental/class-theme-json-webfonts-handler.php';
 require __DIR__ . '/experimental/class-wp-theme-json-resolver-gutenberg.php';
+
+/*
+ * DO NOT MERGE THIS CHANGE.
+ * Don't load the Webfonts API.
+require __DIR__ . '/experimental/register-webfonts-from-theme-json.php';
 require __DIR__ . '/experimental/class-wp-webfonts.php';
 require __DIR__ . '/experimental/class-wp-webfonts-provider.php';
 require __DIR__ . '/experimental/class-wp-webfonts-provider-local.php';
 require __DIR__ . '/experimental/webfonts.php';
+ */
```
This is a temporary workaround for testers to test this PR. Once the handler is backported to Core, then (in a separate PR), `remove_action( 'plugins_loaded', '_wp_theme_json_webfonts_handler' );` can be added to turn off the stopgate handler in Core and use the Webfonts API.

hellofromtonya commented:
@peterwilsoncc @desrosj This PR is using the concepts we talked about. What do you think?

@kjellr @jffng I tested it using the [TT2 alternative style variations PR](https://github.com/WordPress/wordpress-develop/pull/2440). Can you please test it too and share feedback?

costdev commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+
+				foreach ( $webfont['src'] as $url ) {
+					// Add data URIs first.
+					if ( 0 === strpos( trim( $url ), 'data:' ) ) {
```
```suggestion
					if ( str_starts_with( trim( $url ), 'data:' ) ) {
```

costdev commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+
+				foreach ( $value as $item ) {
+
+					if ( 0 === strpos( $item['url'], get_site_url() ) ) {
```
```suggestion
					if ( str_starts_with( $item['url'], get_site_url() ) ) {
```

costdev commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+			private $stylesheet_handle = '';
+
+			/**
+			 * Initialize the handler.
```
```suggestion
			 * Initializes the handler.
```

costdev commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+			}
+
+			/**
+			 * Register webfonts within the `theme.json`.
```
```suggestion
			 * Registers webfonts declared in theme.json.
```

Suggestions:
- Change to "Registers" for consistency in the PR and with Core.
- Remove "the" for consistency in the PR.
- Add "decared in" for clarity.
- Officially, markdown syntax is only valid in the body of a docblock, not the title.
- Other docblock titles in the PR do not encapsulate theme.json at all.

costdev commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+			}
+
+			/**
+			 * Gets the webfonts from `theme.json`.
```
```suggestion
			 * Gets the webfonts from theme.json.
```
- Officially, markdown syntax is only valid in the body of a docblock, not the title.
- Other docblock titles in the PR do not encapsulate theme.json at all.

costdev commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+			}
+
+			/**
+			 * Transform each `src` into an URI by replacing the `file:./`
```
```suggestion
			 * Transforms each 'src' into an URI by replacing 'file:./'
```
- Officially, markdown syntax is only valid in the body of a docblock, not the title.

costdev commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+			 * Transform each `src` into an URI by replacing the `file:./`
+			 * placeholder from theme.json.
+			 *
+			 * The absolute path to the webfont file(s) Cannot be defined in
```
```suggestion
			 * The absolute path to the webfont file(s) cannot be defined in
```

costdev commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+	add_action( 'plugins_loaded', '_wp_theme_json_webfonts_handler' );
+	/**
+	 * Runs the theme.json webfonts handler to expose defined webfonts
+	 * to the editor's typography pickers and generate `@font-face` style
```
```suggestion
	 * to the editor's typography pickers and generate '@font-face' style
```
- Officially, markdown syntax is only valid in the body of a docblock, not the title.

costdev commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+			}
+
+			/**
+			 * Converts the face-face properties (i.e. keys) into kebab-case.
```
```suggestion
			 * Converts the font-face properties (i.e. keys) into kebab-case.
```
- Fonty McFontFace

costdev commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+			}
+
+			/**
+			 * Validate a webfont.
```
```suggestion
			 * Validates a webfont.
```

costdev commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+					return false;
+				}
+
+				// Make a `src` property is defined.
```
```suggestion
				// Check that the `src` property is defined and a valid type.
```

costdev commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+					return false;
+				}
+
+				// Validate the 'src' property.
```
```suggestion
				// Validate the `src` property.
```

For consistency with the block above.

costdev commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+				}
+
+				// Make a `src` property is defined.
+				if ( empty( $webfont['src'] ) || ( ! is_string( $webfont['src'] ) && ! is_array( $webfont['src'] ) ) ) {
```
Since this is run through `wp_parse_args()`, a `src` key should always exist.

However, the value may be falsey. Is a falsey check more appropriate here?

Should this have an additional condition for a string containing only spaces?

```php
|| ( is_string( $webfont['src'] ) && '' === trim( $webfont['src'] ) )
```

costdev commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+				}
+
+				// Validate the 'src' property.
+				if ( ! empty( $webfont['src'] ) ) {
```
With the check above returning `false`, is this check necessary?

costdev commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+				// Validate the 'src' property.
+				if ( ! empty( $webfont['src'] ) ) {
+					foreach ( (array) $webfont['src'] as $src ) {
+						if ( empty( $src ) || ! is_string( $src ) ) {
```
See a related comment above about checking for a value that only contains spaces.

costdev commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+			}
+
+			/**
+			 * Generate and enqueue webfonts styles.
```
```suggestion
			 * Generates and enqueues webfonts styles.
```

costdev commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+			}
+
+			/**
+			 * Generate and enqueue editor styles.
```
```suggestion
			 * Generates and enqueues editor styles.
```

costdev commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+			}
+
+			/**
+			 * Gets the `@font-face` CSS styles for locally-hosted font files.
```
```suggestion
			 * Gets the '@font-face' CSS styles for locally-hosted font files.
```
- Officially, markdown syntax is only valid in the body of a docblock, not the title.

costdev commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+			}
+
+			/**
+			 * Order `src` items to optimize for browser support.
```
```suggestion
			 * Orders 'src' items to optimize for browser support.
```
- Officially, markdown syntax is only valid in the body of a docblock, not the title.

costdev commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+			}
+
+			/**
+			 * Compiles the `src` into valid CSS.
```
```suggestion
			 * Compiles the 'src' into valid CSS.
```
- Officially, markdown syntax is only valid in the body of a docblock, not the title.

costdev commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+			}
+
+			/**
+			 * Add webfonts mime types.
```
```suggestion
			 * Adds webfonts mime types.
```

costdev commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+			 * @since 6.0.0
+			 *
+			 * @param array $mime_types Array of mime types.
+			 * @return array Mime types with webfonts formats.
```
Should have an empty line between the `@param` and `@return` annotations for consistency with the rest of the class.

hellofromtonya replied:
No empty lines should be there, per Core's coding standard.

costdev commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+		}
+
+		/**
+		 * Handler for `theme.json` webfonts.
```
```suggestion
		 * Handler for theme.json webfonts.
```

- Officially, markdown syntax is only valid in the body of a docblock, not the title.
- Other docblock titles in the PR do not encapsulate theme.json at all.

peterwilsoncc commented:
@hellofromtonya I'm getting a fatal error attempting to test this:

```
Fatal error: Uncaught Error: Call to undefined function gutenberg_add_registered_webfonts_to_theme_json() in /vagrant/content/plugins/gutenberg/lib/experimental/class-wp-theme-json-resolver-gutenberg.php on line 39
```

My config is:
* https://github.com/WordPress/wordpress-develop/pull/2440 checked out @ https://github.com/WordPress/wordpress-develop/pull/2440/commits/2a533caa033495f08ff82a7024b6ed6e703f37c9
* Gutenberg active with this branch checked out

I tried excluding `class-wp-theme-json-resolver-gutenberg.php` in the load file but that just caused a different fatal error.

costdev reviewed (changes requested):
Currently unable to test due to internet troubles, so in the meantime, here's some suggestions for docblocks, polyfill usage and additional validation.

I'll get testing when I'm back on a stable connection.

peterwilsoncc commented:
> I'm getting a fatal error attempting to test this:

I resolved this by applying this diff, I think it's correct to limit everything to the stopgap only.

```diff
diff --git a/lib/experimental/class-wp-theme-json-resolver-gutenberg.php b/lib/experimental/class-wp-theme-json-resolver-gutenberg.php
index 960ea659e8..119489e665 100644
--- a/lib/experimental/class-wp-theme-json-resolver-gutenberg.php
+++ b/lib/experimental/class-wp-theme-json-resolver-gutenberg.php
@@ -36,14 +36,14 @@ class WP_Theme_JSON_Resolver_Gutenberg extends WP_Theme_JSON_Resolver_6_0 {
 		if ( null === static::$theme ) {
 			$theme_json_data = static::read_json_file( static::get_file_path_from_theme( 'theme.json' ) );
 			$theme_json_data = static::translate( $theme_json_data, wp_get_theme()->get( 'TextDomain' ) );
-			$theme_json_data = gutenberg_add_registered_webfonts_to_theme_json( $theme_json_data );
+			// $theme_json_data = gutenberg_add_registered_webfonts_to_theme_json( $theme_json_data );
 			static::$theme   = new WP_Theme_JSON_Gutenberg( $theme_json_data );
 
 			if ( wp_get_theme()->parent() ) {
 				// Get parent theme.json.
 				$parent_theme_json_data = static::read_json_file( static::get_file_path_from_theme( 'theme.json', true ) );
 				$parent_theme_json_data = static::translate( $parent_theme_json_data, wp_get_theme()->parent()->get( 'TextDomain' ) );
-				$parent_theme_json_data = gutenberg_add_registered_webfonts_to_theme_json( $parent_theme_json_data );
+				// $parent_theme_json_data = gutenberg_add_registered_webfonts_to_theme_json( $parent_theme_json_data );
 				$parent_theme           = new WP_Theme_JSON_Gutenberg( $parent_theme_json_data );
 
 				// Merge the child theme.json into the parent theme.json.
```

costdev commented:
Tested using the instructions in the PR's description.

Results:
- 2 @font-face declarations to match what was defined in the theme.json file. ✅
- The url should no longer have the file:./ placeholder in it. ✅

Will continue testing.

---
For others testing by making the manual changes to `functions.php` and `theme.json` in TT2, note that you may get a 404 error in console for the font files.

If so, move:

```
assets/fonts/SourceSerif4Variable-Roman.ttf.woff2 (and other SourceSerif font files)
```
to
```
assets/fonts/source-serif-pro/SourceSerif4Variable-Roman.ttf.woff2
```

peterwilsoncc commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+
+					if ( 0 === strpos( $item['url'], get_site_url() ) ) {
+						$item['url'] = wp_make_link_relative( $item['url'] );
+					}
```
```suggestion
					if (
						str_starts_with( $item['url'], site_url() )
						|| str_starts_with( $item['url'], home_url() )
					) {
						$item['url'] = wp_make_link_relative( $item['url'] );
					}
```

It's possible themes can live outside of the `site_url` if the config is something like:
* Home URL: example.com
* Site URL: example.com/wp
* Content URL: example.com/content

I'm suggesting checking for both as the site and home urls can technically be on different domains. 

The shorter function names also return so that's eight bytes that can be saved by dropping the gets

hellofromtonya replied:
Done https://github.com/WordPress/gutenberg/pull/40493/commits/66926e28c600e9eabb3701ec85e9af27bde0b129

peterwilsoncc commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+				// Add woff2.
+				if ( ! empty( $src['woff2'] ) ) {
+					$src_ordered[] = array(
+						'url'    => $src['woff2'],
```
```suggestion
						'url'    => sanitize_url( $src['woff2'] ),
```

Possibly `esc_url()` but I think CSS and HTML escape URLs differently.

I think `file` protocol has been normalized to an actual URL at this point, they'll get stripped if not. data URIs can't be sanitized by core as far as I know.

:1234: Applies to this and the items below.

hellofromtonya replied:
Resolved https://github.com/WordPress/gutenberg/pull/40493/commits/4f6f098598f99a14a400920847f7dc2666c70043

costdev commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+
+					$webfont = $this->convert_keys_to_kebab_case( $webfont );
+
+					$webfont = $this->validate_webfont( $webfont );
+
+					$webfont['src'] = $this->transform_src_into_uri( (array) $webfont['src'] );
```
If `::validate_webfont()` returns `false`, `$webfont['src']` will result in:

```
Notice: Trying to access array offset on value of type bool
```

hellofromtonya replied:
Oh yeah, that's right. Why did I remove that check? 🤦‍♀️  Adding it back in. Thanks.

peterwilsoncc commented on lib/experimental/class-theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,537 @@
+
+				// Register callback to generate and enqueue styles.
+				if ( did_action( 'wp_enqueue_scripts' ) ) {
+					$this->stylesheet_handle = 'wp-webfonts-footer';
+					$hook                    = 'wp_print_footer_scripts';
```
Is the footer fallback needed while only supporting theme.json?

hellofromtonya replied:
Nope. Will review. Good catch!

hellofromtonya replied:
Resolved https://github.com/WordPress/gutenberg/pull/40493/commits/e32700d58e0b97765f1f09a767177389c36d831b.

peterwilsoncc reviewed (commented):
I've added a few notes inline.

I tried forcing my way in to using the API from WP CLI (ie, not in theme.json) and wasn't able to find a reasonable way to do so.

Testing with the TT2 branch of WordPress-Develop shows what I'd expect in the source code.

Commit:
Updates DocBlock from code review.

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

Commit:
Use str_starts_with() instead of strpos().

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

hellofromtonya committed:
Fix guarding conditions.

hellofromtonya committed:
Sanitize src URLs.

hellofromtonya committed:
Remove wp_print_footer_scripts hook.

hellofromtonya commented:
In reviewing with @costdev, an enterprising developer who really wants to gain access could use reflection to get access to within the container's class. This means using a class isn't truly non-accessible. 

I've switched the design to use anonymous functions (closures) that are stored to variables within the container's scope (as originally suggested by @peterwilsoncc).

hellofromtonya committed:
Adds home_url() as a relative link check.

hellofromtonya committed:
Use closures assigned to variables to limit external access.

Changes to use $fn_var = function() {} design pattern, which
eliminates external access. The previous design used a `class`
which when using reflection is not fully unaccessible.

hellofromtonya committed:
Comment out API functions in Resolver for testing stopgap.

hellofromtonya commented:
@peterwilsoncc Thanks for spotting the fatal error. I had commented those 2 lines out in my testing environment but didn't include them in the PR.

Those 2 lines are now commented out in https://github.com/WordPress/gutenberg/pull/40493/commits/fb00786b9e5b52b6e280068208f1bc80a29cb5ab.

jffng reviewed (commented):
👋 

I gave this test using [TT2's style variations](https://github.com/WordPress/wordpress-develop/pull/2440).

The fonts are loading for each variation on the front-end. However in the site editor, only the fonts for the currently saved style variation load:

https://user-images.githubusercontent.com/5375500/164489867-1898d922-d9e1-4d37-a4d8-15afc6abd697.mp4

This issue was at one point fixed [here](https://github.com/WordPress/gutenberg/pull/40262) but I'm not sure if that is still relevant to this PR's implementation.

hellofromtonya committed:
Add webfonts defined in variations when in the editor.

hellofromtonya commented:
Thanks @jffng for pointing that out. I pushed an update https://github.com/WordPress/gutenberg/pull/40493/commits/8295f151b7f7b7389e4b9bccd3204f563bc03931 that brings PR #39886 and #40262 into this PR. Can you retest please?

hellofromtonya commented:
Whoops, sorry I missed the variable name changes. Fixed now in https://github.com/WordPress/gutenberg/pull/40493/commits/8295f151b7f7b7389e4b9bccd3204f563bc03931.

jffng commented:
Getting the following fatal: 

`
Fatal error: Uncaught Error: Call to undefined method WP_Theme_JSON_Resolver::get_style_variations() in /Users/jeffreyong/Documents/THEMES/a8c-projects/gutenberg/lib/experimental/class-theme-json-webfonts-handler.php on line 47
`

hellofromtonya commented:
Hmm, that new method has been merged into Core (in `trunk`) `WP_Theme_JSON_Resolver::get_style_variations()` https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/class-wp-theme-json-resolver.php#L472. I wonder if Gutenberg's Docker image isn't using `trunk` but instead is using 5.9 release.

I'll add a temporary check to load either Core's version or `WP_Theme_JSON_Resolver_6_0` version to accommodate testers who use either testing environment.

hellofromtonya committed:
Use WP_Theme_JSON_Resolver_6_0 instead of latest in WP Core trunk.

hellofromtonya commented:
Hey @jffng, I switched to use Gutenberg's copy of what's in WP Core trunk https://github.com/WordPress/gutenberg/pull/40493/commits/6fb945b84dacab47ddf4fb873640c0272cb46fb3. Can you retest please?

costdev commented:
> Getting the following fatal:
> 
> `Fatal error: Uncaught Error: Call to undefined method WP_Theme_JSON_Resolver::get_style_variations() in /Users/jeffreyong/Documents/THEMES/a8c-projects/gutenberg/lib/experimental/class-theme-json-webfonts-handler.php on line 47`

I got the same error and resolved it by changing the call from `WP_Theme_JSON_Resolver::get_style_variations()` to `WP_Theme_JSON_Resolver_6_0::get_style_variations()`

Edit: Ah, just saw the commit you linked to Tonya 😛

jffng commented:
> I switched to use Gutenberg's copy of what's in WP Core trunk https://github.com/WordPress/gutenberg/commit/6fb945b84dacab47ddf4fb873640c0272cb46fb3. Can you retest please?

That did the trick, thank you! All variations and their respective fonts working in the site editor and front-end for me.

bph commented:
Tested it with TT2 and modifications according to above instructions today's modifications. 

Font face as expected in the header / front end
![Screen Shot 2022-04-21 at 12 18 55](https://user-images.githubusercontent.com/39980/164521818-26a7071d-39f9-4b6c-a67f-2aa9988e7721.png)
Font face as expected in the header / front end
![Screen Shot 2022-04-21 at 18 44 34](https://user-images.githubusercontent.com/39980/164563505-97938654-ff17-4db6-9135-8a1de9cdb9e9.png)

Font Family in dropdown for Typography was available (Styles, Site Title block) 
![Screen Shot 2022-04-21 at 13 56 06](https://user-images.githubusercontent.com/39980/164522081-5563b480-41dc-43b4-b1a5-6d81b8fbc83d.png)

hellofromtonya committed:
Rename file, no longer a class.

hellofromtonya committed:
Relocate file for backport.

hellofromtonya committed:
Remove reassignment of $settings array level init (micro-optimization).

costdev commented:
I've been testing with the design change to use anonymous functions as variables within the container. This appears to fully prevent direct interaction with the API. In other words, I can't find a way to use the API in a backwards-incompatible way.

LGTM 👍

ndiego commented:
Hi all, do we have an ETA on when this will be merged?

hellofromtonya commented:
>Hi all, do we have an ETA on when this will be merged?

@ndiego before Beta 3's (April 26th) release party. Here's the backport PR in WP Core https://github.com/WordPress/wordpress-develop/pull/2622.

hellofromtonya commented:
Closing this PR. Why? 

* It's been backported to WordPress Core for 6.0 Beta 3 via https://core.trac.wordpress.org/changeset/53282. 
* Does not need to exist in Gutenberg.

Thank you everyone for contributing!

oandregal commented on lib/compat/wordpress-6.0/theme-json-webfonts-handler.php:
```diff
@@ -0,0 +1,521 @@
+			}
+
+			// Bail out early if there are no settings for webfonts.
+			if ( empty( $settings['typography'] ) || empty( $settings['typography']['fontFamilies'] ) ) {
```
Font families can be defined at the top-level, as in:

```json
{
  "settings": {
      "typography": {
          "fontFamilies": []
      }
  }
}
```

But also can be defined per block, as in:

```json
{
  "settings": {
      "typography": {
          "fontFamilies": [] // Fonts for all.
      },
      "blocks": {
          "some/block": {
              "typography": {
                  "fontFamilies": [] // A block can define its own.
              }
          }
      }
  }
}
```

Is this problematic for webfonts? How can a webfont be defined for a single block?

hellofromtonya replied:
>Is this problematic for webfonts? How can a webfont be defined for a single block?

@oandregal Top-level only in this first version. 

I remember a discussion thread (not sure where it happened) where @mtias indicated not having block-level web font selection. Also see https://github.com/WordPress/gutenberg/issues/40362.
